File: 2024/2/2.jl
#!/usr/bin/env julia
# usage: julia 2.jl [FILENAME]

module AOC2024Day2

function safe(row)::Bool
    rowdiff = diff(row)
    all(in(1:3), rowdiff) || all(in(-3:-1), rowdiff)
end

load(file) = [parse.(Int, split(line)) for line in eachline(file)]

selectout(list) = ([list[begin:i-1]; list[i+1:end]] for i in eachindex(list))

partone(input) = count(safe, input)
parttwo(input) = count(row -> any(safe, selectout(row)), input)

if abspath(PROGRAM_FILE) == @__FILE__
    input = load(get(ARGS, 1, joinpath(@__DIR__, "input.txt")))
    println(partone(input))
    println(parttwo(input))
end

end # module

File: 2024/2/2.pl
#!/usr/bin/env swipl

:- module(aoc2024day2, []).
:- module(aoc2024day2).
:- initialization(main, main).
:- use_module(library(clpfd)).
:- use_module(library(dcg/basics)).
:- use_module(library(dcg/high_order)).
:- use_module(library(yall)).

load(Rows) --> sequence(row, Rows).
row(Row) --> sequence(integer, " ", Row), { Row \= [] }, eol.

safe(Row) :-
    once((
        member(Range, [1..3, (-3)..(-1)]),
        forall(nextto(X, Y, Row), (Z #= X - Y, Z in Range))
    )).

part_one(Rows, Result) :-
    aggregate_all(count, (member(Row, Rows), safe(Row)), Result).

part_two(Rows, Result) :-
    aggregate_all(count, (
        member(Row, Rows), once(( select(_, Row, Row0), safe(Row0) ))
    ), Result).

main([File]) :-
    once(phrase_from_file(load(Input), File)),
    part_one(Input, ResultOne),
    writeln(ResultOne),
    part_two(Input, ResultTwo),
    writeln(ResultTwo).

main([]) :- main(['input.txt']).
